### PR TITLE
Update Inspect bg plot when parameters changed in EngDiff UI fitting tab

### LINKED ISSUE #33246: Inspect background plot in EngDiff UI not updating when parameters changed in table
**Describe the bug**
Found during manual testing for 6.3 release - #33191
Inspect background plot inn EngDiff UI not updating when parameters (for EnggEstimateFocusedBackground) changed in UI table

**To Reproduce**
(1) Load in a focused spectrum into the fitting tab of EngDiff UI spectrum
(2) Click any cell in the UI table to select the row
(3) Click Inspect Background button (a separate figure window should open)
(4) Zoom in on the Red curve in the top subplot (this is the estimated background)
(5) In the table change the Niter column to be 1 - the red curve will not have changed (but it should have!)
(6) Click Inspect Background button again - a new plot should open and the background curve should have more prominent peaks in it where there are peaks in the data.

**Expected behavior**
The background curve in the top subplot (and the subtracted data in the bottom subplot) should update when these parameters are changed (NIter, XWindow, SG).

**Platform/Version (please complete the following information):**
 - Nightly 6.2 (4 Jan) windows

### LINKED ISSUE #33247: Add legend to Inspect background plot in EngDiff UI
**Describe the bug**
Found during manual testing for 6.3 release - #33191 
Add legend to Inspect background plot in EngDiff UI

**To Reproduce**
(1) Load in a focused spectrum into the fitting tab of EngDiff UI spectrum
(2) Click any cell in the UI table to select the row
(3) Click Inspect Background button (a separate figure window should open)

**Expected behavior**
Would have a legend

### PR DESCRIPTION
**Pull request by RichardWaiteSTFC**
**Description of work.**

The background estimated for a spectrum is not tracked in the ADS, this did meant that when the background estimation parameters (passed to `EnggEstimateFocusedBackground`) were changed the background curve in the Inspect Background plot wasn't updated (the background subtracted data however were updated on the plot).

This problem could have been fixed by storing the bg in the ADS with a  unique name, however I think it is good to limit the number of workspaces present in the ADS (though it could always be hidden) and not add another set of workspaces to keep track of when workspaces are renamed/deleted etc. 

This change adds in a function that updates the background data when the figure is redrawn. I have also added a check that there is a tracked and untracked workspace present (data and background curve respectively) before trying to update the data on the background curve, I could add more checks (e.g. same number of data points, correct labels etc.) but I think it's overkill as it's an external plot that can always be regenerated with one click if the user gets it into an unusable state.

Have also added a legend to make it clearer which curve is which (fixes #33247).

**To test:**

(1) Follow instructions in issue
(2) Load two spectra and open the Inspect background figure for each, check the correct plot is updated when a bg param is changed
(3) Try to remove the data or background curve from one of the figures (cog > Curves > Remove) it should print a warning to the log `Inspect background figure 1 has been invalidated - the background curve will no longer be updated.`
(4) Update the bg params in the table (e.g. Niter=1) - the background curve should not update (the same warning will also be printed)
(5) Delete the bgsub workspace corresponding to the data on the remaining figure, it should print the same warning.

Fixes #33246
Fixes #33247

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/fitting/data_handling/data_model.py [hunk 1]
```diff
@@ -144,6 +144,7 @@ def __init__(self):
         self._fit_workspaces = None
         self._last_added = []  # List of workspace names loaded in the last load action.
         self._data_workspaces = FittingWorkspaceRecordContainer()
+        self.inspect_bg_fig = None
```
Comment (RichardWaiteSTFC) [inline qt/python/mantidqtinterfaces/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/fitting/data_handling/data_model.py:147]: Sorry no longer need this - will remove it!

**Review by DanielMurphy22 — CHANGES_REQUESTED** [review]
Trying to inspect background I get an error with the legend. Possilby a linux only problem:

```
Traceback (most recent call last):
  File "/opt/mantidunstable/lib/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/fitting/data_handling/data_presenter.py", line 179, in _plotBG
    self.model.plot_background_figure(ws_name)
  File "/opt/mantidunstable/lib/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/fitting/data_handling/data_model.py", line 543, in plot_background_figure
    ax[0].legend(fontsize=8.0).set_draggable(True).legend
AttributeError: 'Legend' object has no attribute 'set_draggable'
```

**Review by DanielMurphy22 — DISMISSED** [review]
Tested on Ubuntu: Simply inspecting the background and trying to do step 2 and 3  are really slow. Going to double check on IDAaaS as I understand that's not overly informative advice!

I get this error which I think is a separate issue. Maybe these plots shouldn't be project saved as they are not in the ads??
```
Python-[Debug] Project Recovery: Saving started
Python-[Debug] Plot: 1 was not saved. Error: 'label1On'
Python-[Debug] Plot: 2 was not saved. Error: 'label1On'
Python-[Debug] Project Recovery: Saving finished
```

**Comment by RichardWaiteSTFC** [comment]
> Tested on Ubuntu: Simply inspecting the background and trying to do step 2 and 3 are really slow. Going to double check on IDAaaS as I understand that's not overly informative advice!
> 
> I get this error which I think is a separate issue. Maybe these plots shouldn't be project saved as they are not in the ads??
> 
> ```
> Python-[Debug] Project Recovery: Saving started
> Python-[Debug] Plot: 1 was not saved. Error: 'label1On'
> Python-[Debug] Plot: 2 was not saved. Error: 'label1On'
> Python-[Debug] Project Recovery: Saving finished
> ```

Not noticed any slow down myself on windows, but neither have I seen that error before. There is no need to save these plots in project save so I will see if they can be excluded and if so hopefully that will fix the error and speed it up!

**Comment by RichardWaiteSTFC** [comment]
Actually @DanielMurphy22 - on second thoughts the slow down might be due to recursive calls to the on_draw handle - could you add a print statement there to confirm this?

**Review by DanielMurphy22 — COMMENTED** [review]
Testing on Rhel7 (IDAaaS development instance).  It worked really well, I'd like someone else to check if they see the slow down on Ubuntu, but otherwise I'd be happy to approve! 

To test the slow down, build this branch on Ubuntu 18.04 and load data into the fitting tab , inspect background on both and try to tick/untick SG in the table and remove a curve in the figure options of the bakground plots. This slow down will be obvious if you see it!!

Just a few notes:
- Not a problem, but can open an inspect background plot on the same data multiple times
- I've nothing plotted in the fitting tab but making changes to the table replots my empty docked plot!!
- Project recovery point above!

**Comment by DanielMurphy22** [comment]
> 

I'll try! as i said it seemed fine on idaaas so maybe once packaged this is fine??

**Comment by DanielMurphy22** [comment]
> > Tested on Ubuntu: Simply inspecting the background and trying to do step 2 and 3 are really slow. Going to double check on IDAaaS as I understand that's not overly informative advice!
> > I get this error which I think is a separate issue. Maybe these plots shouldn't be project saved as they are not in the ads??
> > ```
> > Python-[Debug] Project Recovery: Saving started
> > Python-[Debug] Plot: 1 was not saved. Error: 'label1On'
> > Python-[Debug] Plot: 2 was not saved. Error: 'label1On'
> > Python-[Debug] Project Recovery: Saving finished
> > ```
> 
> Not noticed any slow down myself on windows, but neither have I seen that error before. There is no need to save these plots in project save so I will see if they can be excluded and if so hopefully that will fix the error and speed it up!

Also disabled project recovery!

**Comment by DanielMurphy22** [comment]
> Actually @DanielMurphy22 - on second thoughts the slow down might be due to recursive calls to the on_draw handle - could you add a print statement there to confirm this?

printed about 100 times / second non stop after opening the plot on Ubuntu! BUT on idaaas only prints twice anytime I make a genuine change to the table (so works wonderfully). No idea whats wrong, but it's this call!

**Comment by RichardWaiteSTFC** [comment]
> > Actually @DanielMurphy22 - on second thoughts the slow down might be due to recursive calls to the on_draw handle - could you add a print statement there to confirm this?
> 
> printed about 100 times / second non stop after opening the plot on Ubuntu! BUT on idaaas only prints twice anytime I make a genuine change to the table (so works wonderfully). No idea whats wrong, but it's this call!
> 
> BTW: I've made the print statement the first line of on_draw().

OK so i think it's because ubuntu must have a different version of matplotlib and/or a different back-end. I will try to add some `max_recursion_depth`  or something!

**Comment by RichardWaiteSTFC** [comment]
Instead of checking that the bg points are the same I instead modify a bool property of the canvas to indicate whether the bg has been drawn - the property `signalsBlocked` is harmless for the duration of the draw loop - though I had first intended it to stop the `on_draw` being called when updating the background (it does not). This fixes the slowdown on ubuntu.

With regard to the other issues found I do not think these are regressions.
To summarise our discussion offline:

1. Project recovery plot saver error
```
Python-[Debug] Project Recovery: Saving started
Python-[Debug] Plot: 1 was not saved. Error: 'label1On'
```
I think this might be a more general problem that mantid can't save plots with non ADS data e.g.
```
# i) make plot
fig, ax = plt.subplots(subplot_kw={'projection': 'mantid'})  # same error without projection
ax.plot([1,2,3])
fig.show()
# ii)  Save the project (File > Save Project As)
# => Plot: 2 was not saved. Error: list index out of range
```
Different error but the end result is the plot isn't saved.
The rest of the project is however saved and in the case of the EngDiff UI appears to be functioning. So I think this is not an issue for us as these inspect bg plots are used for diagnostic purposes only and not intended to be saved/restored - it can be generated at the click of a button anyway.

2. Multiple inspect background plots can be opened on the same data (though all update as expected with this fix). As discussed this is because the plots and model don't know about each other and this could be solved by storing the figure number in a dict with the workspace name in the model. However this dict needs to be kept updated with ADS delete/clear/rename etc. so this is probably a low priority. If you're happy to approve as is, then I will make an issue for this.

**Review by DanielMurphy22 — APPROVED** [review]
This works really well for me on Ubuntu now!! Good job!

**Comment by sf1919** [comment]
Does this need a release note?

**Comment by RichardWaiteSTFC** [comment]
> Does this need a release note?

I think it's a regression so is fine without a release note

**Pull request merged**